`FreeSpace` allocation metric impl

dgrisham opened the pull request:
Currently a WIP, but looking for feedback on implementation. I've added a `disk.NewInformerWithMetric(metric string)` function that supports `"disk-reposize"` and `"disk-freespace"` metrics (values checked in `setupAllocation()`, may want to check in `disk.NewInformerWithMetric()` instead). I think the disk Informer impl is fairly clean and makes it easy to add new metrics (simply by adding an entry to the `nameToRPC` map), but let me know what you think @hsanjuan.

TODO:
- [x] implement `descendalloc`
- [x] Tests for freespace metric

hsanjuan commented:
I think there are files missing (util)

dgrisham commented:
@hsanjuan Thanks, was in a hurry when I pushed. Just added the missing files

hsanjuan commented on allocator/allocator.go:
```diff
@@ -0,0 +1,91 @@
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+type Allocator interface {
```
Dup from the PinAllocator interface? I'm not sure why it's needed

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -12,13 +12,21 @@ import (
 	"github.com/ipfs/ipfs-cluster/api"
 )
 
+// TODO: switch default to disk-freespace
```
dont forget

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -28,6 +36,14 @@ type Informer struct {
 
 // NewInformer returns an initialized Informer.
 func NewInformer() *Informer {
+	MetricName = DefaultMetric
```
On the line of previous comment, I think you can simply make the metric name a property of the informer...

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -12,13 +12,21 @@ import (
 
 // MetricName specifies the name of our metric
-var MetricName = "disk"
+var MetricName string
```
Before this was more like a configuration option for the module, but now it's a property of the informer defined on creation, so it should not exist anymore as global variable.

dgrisham replied:
Ah, okay I see. I was a bit uncertain if I should change that or not, your explanation makes sense though.

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -49,30 +65,28 @@ func (disk *Informer) Name() string {
 	err := disk.rpcClient.Call("",
 		"Cluster",
-		"IPFSRepoSize",
+		nameToRPC[MetricName],
```
either handle the metricName -> rpcname conversion via a switch statement (doing everything here in the function) with a `default` section or add code to make sure that the metric name is in the nameToRPC map. If not, log the problem clearly and return an invalid metric.

Right now the error logged would probably be very cryptic (rpc method missing i guess) and probably difficult to figure out it's just from a typo in the metric name.

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -12,13 +12,21 @@ import (
 	"github.com/ipfs/ipfs-cluster/api"
```
I can't place comment before this line: The godoc description of the module needs to be updated to explain the things this informer disk informer can do.

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -28,6 +36,14 @@ type Informer struct {
 
 // NewInformer returns an initialized Informer.
```
mention that it will use the default metric

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -28,6 +36,14 @@ type Informer struct {
+	return &Informer{}
+}
+
+// NewInformer returns an initialized Informer.
```
mention which are the valid metrics that can be used

hsanjuan commented on allocator/allocator.go:
```diff
@@ -0,0 +1,91 @@
+func (alloc SimpleAllocator) Shutdown() error { return nil }
+
+// Allocate does nothing in this allocator (important to override this)
+func Allocate(c *cid.Cid, current, candidates map[peer.ID]api.Metric) ([]peer.ID, error) {
```
I think this was meant to be  `func (alloc SimpleAllocator) Allocate()?`

hsanjuan commented on allocator/allocator.go:
```diff
@@ -0,0 +1,91 @@
+
+// SimpleAllocator implements the Allocator interface. Other allocators can
+// extend SimpleAllocator and should at least override the Allocate() function.
+type SimpleAllocator struct{}
```
I'm not very sure why a `SimpleAllocator` which does nothing is necessary...

hsanjuan commented on allocator/allocator.go:
```diff
@@ -0,0 +1,91 @@
+
+// MetricsSorter attaches sort.Interface methods to our metrics and sorts
+// a slice of peers in the way that interest us
+type MetricsSorter interface {
```
hmm this is just re-defining https://golang.org/pkg/sort/#Interface

dgrisham replied:
Right, I realized that before but then apparently forgot it completely while playing with Go's type system.

hsanjuan commented on allocator/allocator.go:
```diff
@@ -0,0 +1,91 @@
+
+// SimpleSorter implements the MetricsSorter interface. Other sorters may extend
+// this and should at least override the default Less() function.
+type SimpleSorter struct {
```
Probably remove the MetricSorter interface and rename this to MetricSorter (a type that implements the sort.Interface for metrics). The type can have a property `Reverse` which would allow it to sort in reverse order.

hsanjuan commented on allocator/ascendalloc/ascendalloc.go:
```diff
@@ -1,96 +1,50 @@
-// NewAllocator returns an initialized Allocator
-func NewAllocator() *Allocator {
-	return &Allocator{}
+// AscendAllocator extends the SimpleAllocator
```
SimpleAllocator does nothing so I don't quite get the benefit of extending. I think it is simpler to implement the ipfscluster.PinAllocator interface directly and use the common MetricSorter.

dgrisham commented on ipfs-cluster-service/main.go:
```diff
@@ -342,8 +344,12 @@ func setupLogging(lvl string) {
-		return disk.NewInformer(), ascendalloc.NewAllocator()
+	case "disk", "disk-freespace":
+		informer := disk.NewInformer()
+		return informer, descendalloc.NewAllocator()
```
@hsanjuan It might be nice to have a way for the `Informer` to specify which `Allocator` it should use based on the metric (not sure whether there's a good reason to keep those decoupled like they are now).

hsanjuan replied:
They're meant to be fully decoupled. You might gather metrics for a bunch of  reasons not related to an allocator. Then, there might be metrics which make sense with several allocators.

dgrisham replied:
Gotcha, that makes sense.

hsanjuan commented:
Tests related to metric/allocator are failing...

hsanjuan commented on cluster_test.go:
```diff
@@ -79,6 +79,7 @@ func (ipfs *mockConnector) PinLs(filter string) (map[string]api.IPFSPinStatus, e
 
 func (ipfs *mockConnector) ConnectSwarms() error                          { return nil }
 func (ipfs *mockConnector) ConfigKey(keypath string) (interface{}, error) { return nil, nil }
+func (ipfs *mockConnector) FreeSpace() (int, error)                       { return 0, nil }
```
Probably this makes the tests fail, as you report 0 free space, nothing can be allocated

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -12,23 +13,44 @@ import (
+// the nameToRPC map.
+func NewInformerWithMetric(metric string) *Informer {
+	// check whether specified metric is supported
+	if _, valid := nameToRPC[metric]; valid {
```
There something here I don't fully like.

I think we need constants for allowed metric types for the disk informer:

```go
const (
  MetricFreeSpace = iota
  MetricRepoSize
)
```

`NewInformerWithMetric(metric int)` should panic and inform the user that given metric is invalid. It should also log a critical error (try to see if logger.Critical works, otherwise logger.Error will do).

`nameToRPC` becomes a map from int -> string

`Informer` can store directly the RPC name (that way there is no need to read it from the map on every GetMetric).

hsanjuan replied:
Actually `type MetricType int` with NewInformerWithMetric(metric MetricType)`

hsanjuan commented on allocator/descendalloc/descendalloc_test.go:
```diff
@@ -0,0 +1,116 @@
+	{ // regular sort
+		candidates: map[peer.ID]api.Metric{
+			peer0: api.Metric{
+				Name:   numpin.MetricName,
```
Change numpin.Metric name to DefaultMetric or something.. this was probably an old copypasta bug, with the side effect that it imports a totally unrelated module

dgrisham commented:
@hsanjuan Just made an intermediate push addressing some of your review, but not all of it. As of now, I'm treating the `Informer`'s name as completely separate from the `MetricType` -- the `Type` (`MetricType`) determines how the `Informer` behaves, and the `name` is simply the user-facing name used to identify the `Informer` (so it can be anything).

dgrisham commented:
Hm, lot of failed tests on that last round. I'll look at this closer soon.

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -12,23 +13,57 @@ import (
+	}
+}
+
+func NewInformerWithMetric(metric MetricType, name string) *Informer {
```
godoc comment missing

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -44,35 +79,28 @@ func (disk *Informer) Shutdown() error {
 	m := api.Metric{
-		Name:  MetricName,
-		Value: fmt.Sprintf("%d", repoSize),
+		Name:  disk.rpcName,
```
disk.name is perhaps better?

hsanjuan commented on informer/disk/disk.go:
```diff
@@ -12,23 +13,61 @@ import (
+
+// NewInformerWithMetric returns an Informer that uses the input MetricType. The
+// name argument has the same purpose as in NewInformer.
+func NewInformerWithMetric(metric MetricType, name string) *Informer {
```
Can we return (*Informer, error) and set the error of the metric type is invalid here?

dgrisham replied:
Ah, right, good idea. Definitely want to do that.

hsanjuan commented:
@dgrisham ops I didn't see your last comment on irc about wanting to debug the issue yourself, so I went ahead and fixed it. I found it pretty quickly and it was something silly (replace ascendalloc with descendalloc in cluster initialization).

Went over things again and left one last comment. I think I can merge after. Good job!

dgrisham commented:
@hsanjuan No worries, glad it was relatively simple -- figured it had to be. I'll make that change and push it up!

dgrisham commented:
@hsanjuan Hm, so right now the default metric used within cluster is determined within `setupAllocation()`, and the `NewInformer()` method is never used -- do you think it's worth keeping the `NewInformer()` separate from `NewInformerWithMetric()`? The way I see it, we could handle the default metric in two ways:

1. Initialize with the name `"disk"`, like `NewInformer("disk")`. This would indicate that the user was using the default `"disk"` metric, but would not clarify which one.
2. Initialize with a specific name, like "`disk-freespace"` (which is what we're doing now). In this case we never call `NewInformer()`, which I think makes sense because if we did it could be confusing -- if we initialize with `NewInformer("disk-freespace")` but later the `DefaultMetric` changes to something else, then we might forget to change the name in `NewInformer("disk-freespace")` which could result in something confusing/difficult to debug.

The way we're doing it now seems best to me, but then I'm not sure it makes sense to keep a `NewInformer()` that uses a default metric.

dgrisham commented:
@hsanjuan Decided to keep `NewInformer` but remove the `name` argument (automatically set to to `"disk-default"`).

hsanjuan commented:
THANKS @dgrisham !
